File: bot/appearance.py
from __future__ import annotations

import re
from typing import Any

from openrouter_client import chat_completion

CHARACTER_SYSTEM = (
    "You write extremely detailed physical appearance descriptions for roleplay characters. "
    "Describe the person in the image in second person, beginning with the words 'You are'. "
    "Be exhaustive: face shape and features, eyes, eyebrows, nose, lips, skin tone and texture, "
    "hair color/style/length, body type and build, height impression, posture, clothing and fabrics, "
    "colors, accessories, jewelry, expression, mood, art style if illustrated, lighting, and any "
    "distinguishing marks. Aim for roughly 1000–1500 tokens. Write only the description, no preamble."
)

KNOWN_USER_SYSTEM = (
    "You write extremely detailed physical appearance descriptions for roleplay. "
    "Describe the person in the image in third person. Your output must begin with the lowercase "
    "word 'is ' (as it will follow a name like 'Alice (user:123) is ...'). "
    "Be exhaustive: face, hair, body, clothing, colors, expression, art style, and distinguishing "
    "details. Aim for roughly 1000–1500 tokens. Write only the description starting with 'is ', "
    "no preamble."
)

DISCORD_BATCH_SYSTEM = (
    "You write extremely detailed descriptions of images for roleplay context. "
    "For each image, describe everything visible in exhaustive detail: people, objects, places, "
    "colors, composition, mood, text if any, and art style. Aim for roughly 1000–1500 tokens "
    "per image. Output using exactly this format for each image:\n"
    "---IMAGE N---\n"
    "(description)\n"
    "Use N = 1, 2, 3, or 4 matching the image order given."
)


def _image_part(data_url: str) -> dict[str, Any]:
    return {"type": "image_url", "image_url": {"url": data_url}}


async def describe_character_portrait(
    *,
    api_key: str,
    owner_discord_id: str,
    data_url: str,
) -> str:
    content: list[dict[str, Any]] = [
        {"type": "text", "text": "Describe this character portrait."},
        _image_part(data_url),
    ]
    return await chat_completion(
        api_key=api_key,
        owner_discord_id=owner_discord_id,
        system=CHARACTER_SYSTEM,
        user_content=content,
        use_vision=True,
        max_tokens=2000,
    )


async def describe_known_user_portrait(
    *,
    api_key: str,
    owner_discord_id: str,
    data_url: str,
) -> str:
    content: list[dict[str, Any]] = [
        {"type": "text", "text": "Describe this person's appearance."},
        _image_part(data_url),
    ]
    text = await chat_completion(
        api_key=api_key,
        owner_discord_id=owner_discord_id,
        system=KNOWN_USER_SYSTEM,
        user_content=content,
        use_vision=True,
        max_tokens=2000,
    )
    if text and not text.lower().startswith("is "):
        text = f"is {text.lstrip()}"
    return text


def parse_discord_batch_descriptions(raw: str, expected: int) -> list[str]:
    parts = re.split(r"---IMAGE\s+(\d+)---", raw, flags=re.IGNORECASE)
    by_index: dict[int, str] = {}
    i = 1
    while i + 1 < len(parts):
        try:
            idx = int(parts[i])
        except ValueError:
            i += 2
            continue
        body = parts[i + 1].strip()
        if body:
            by_index[idx] = body
        i += 2
    if by_index:
        return [by_index[n] for n in sorted(by_index) if n in by_index][:expected]

    chunks = [c.strip() for c in raw.split("\n\n") if c.strip()]
    if len(chunks) >= expected:
        return chunks[:expected]
    if chunks:
        return chunks
    return [raw.strip()] if raw.strip() else []


async def describe_discord_message_images(
    *,
    api_key: str,
    owner_discord_id: str,
    data_urls: list[str],
) -> list[str]:
    if not data_urls:
        return []
    urls = data_urls[:4]
    content: list[dict[str, Any]] = [
        {
            "type": "text",
            "text": (
                f"Describe each of the following {len(urls)} image(s) in order. "
                "Use ---IMAGE N--- headers as instructed."
            ),
        }
    ]
    for index, url in enumerate(urls, start=1):
        content.append({"type": "text", "text": f"Image {index}:"})
        content.append(_image_part(url))

    raw = await chat_completion(
        api_key=api_key,
        owner_discord_id=owner_discord_id,
        system=DISCORD_BATCH_SYSTEM,
        user_content=content,
        use_vision=True,
        max_tokens=min(8000, 2000 * len(urls)),
    )
    parsed = parse_discord_batch_descriptions(raw, len(urls))
    while len(parsed) < len(urls):
        parsed.append("")
    return parsed[: len(urls)]


async def ensure_character_appearance(
    character: dict,
    owner_discord_id: str,
    api_key: str,
) -> dict:
    if not character.get("imageS3Key") or character.get("appearance"):
        return character
    from dynamo import update_character_fields
    from s3_images import fetch_image_data_url

    data_url = fetch_image_data_url(
        character.get("imageS3Key", ""),
        character.get("imageContentType"),
    )
    if not data_url:
        return character
    slug = character.get("slug") or character["sk"].split("#CHAR#")[-1].split("#")[0]
    appearance = await describe_character_portrait(
        api_key=api_key,
        owner_discord_id=owner_discord_id,
        data_url=data_url,
    )
    if appearance:
        update_character_fields(owner_discord_id, slug, {"appearance": appearance})
        character = dict(character)
        character["appearance"] = appearance
    return character


async def ensure_known_user_appearance(
    known_user: dict,
    owner_discord_id: str,
    slug: str,
    api_key: str,
) -> dict:
    if not known_user.get("imageS3Key") or known_user.get("appearance"):
        return known_user
    from dynamo import update_item_fields
    from s3_images import fetch_image_data_url

    data_url = fetch_image_data_url(
        known_user.get("imageS3Key", ""),
        known_user.get("imageContentType"),
    )
    if not data_url:
        return known_user
    appearance = await describe_known_user_portrait(
        api_key=api_key,
        owner_discord_id=owner_discord_id,
        data_url=data_url,
    )
    if appearance:
        ku_id = known_user.get("knownUserId") or known_user["sk"].split("#KNOWN#")[-1]
        sk = f"USERID#{owner_discord_id}#CHAR#{slug}#KNOWN#{ku_id}"
        update_item_fields("USERS", sk, {"appearance": appearance})
        known_user = dict(known_user)
        known_user["appearance"] = appearance
    return known_user

File: web/src/lib/appearance.ts
import { decryptApiKey } from "./crypto";
import { VISION_MODEL } from "./models";
import { updateItem, userSk } from "./dynamo";

const CHARACTER_SYSTEM =
  "You write extremely detailed physical appearance descriptions for roleplay characters. " +
  "Describe the person in the image in second person, beginning with the words 'You are'. " +
  "Be exhaustive: face shape and features, eyes, eyebrows, nose, lips, skin tone and texture, " +
  "hair color/style/length, body type and build, height impression, posture, clothing and fabrics, " +
  "colors, accessories, jewelry, expression, mood, art style if illustrated, lighting, and any " +
  "distinguishing marks. Aim for roughly 1000–1500 tokens. Write only the description, no preamble.";

const KNOWN_USER_SYSTEM =
  "You write extremely detailed physical appearance descriptions for roleplay. " +
  "Describe the person in the image in third person. Your output must begin with the lowercase " +
  "word 'is ' (as it will follow a name like 'Alice (user:123) is ...'). " +
  "Be exhaustive: face, hair, body, clothing, colors, expression, art style, and distinguishing " +
  "details. Aim for roughly 1000–1500 tokens. Write only the description starting with 'is ', " +
  "no preamble.";

const BILLING_MULTIPLIER = 2;

type ContentPart =
  | { type: "text"; text: string }
  | { type: "image_url"; image_url: { url: string } };

type VisionUsage = {
  text: string;
  inputTokens: number;
  outputTokens: number;
  costUsd: number;
  cachedTokens: number;
  cacheWriteTokens: number;
};

async function visionCompletion(
  apiKey: string,
  system: string,
  content: ContentPart[],
  maxTokens: number,
): Promise<VisionUsage> {
  const resp = await fetch("https://openrouter.ai/api/v1/chat/completions", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      model: VISION_MODEL,
      max_tokens: maxTokens,
      messages: [
        { role: "system", content: system },
        { role: "user", content },
      ],
    }),
  });
  if (!resp.ok) {
    const err = await resp.text();
    throw new Error(`OpenRouter vision failed: ${err}`);
  }
  const data = (await resp.json()) as {
    choices?: { message?: { content?: string } }[];
    usage?: {
      prompt_tokens?: number;
      completion_tokens?: number;
      cost?: number;
      prompt_tokens_details?: { cached_tokens?: number; cache_write_tokens?: number };
    };
  };
  const usage = data.usage ?? {};
  const details = usage.prompt_tokens_details ?? {};
  const rawCost = typeof usage.cost === "number" ? usage.cost : 0;
  return {
    text: (data.choices?.[0]?.message?.content ?? "").trim(),
    inputTokens: usage.prompt_tokens ?? 0,
    outputTokens: usage.completion_tokens ?? 0,
    costUsd: BILLING_MULTIPLIER * rawCost,
    cachedTokens: details.cached_tokens ?? 0,
    cacheWriteTokens: details.cache_write_tokens ?? 0,
  };
}

export async function updateUsage(
  discordId: string,
  inputTokens: number,
  outputTokens: number,
  costUsd: number = 0,
  cachedTokens: number = 0,
  cacheWriteTokens: number = 0,
) {
  await updateItem(
    "USERS",
    userSk(discordId),
    "ADD usageInputTokens :i, usageOutputTokens :o, usageCostUsd :c, usageCachedTokens :ct, usageCacheWriteTokens :cw",
    {
      ":i": inputTokens,
      ":o": outputTokens,
      ":c": costUsd,
      ":ct": cachedTokens,
      ":cw": cacheWriteTokens,
    },
  );
}

function imagePart(dataUrl: string): ContentPart {
  return { type: "image_url", image_url: { url: dataUrl } };
}

export async function describeCharacterPortrait(
  ownerDiscordId: string,
  encryptedApiKey: string,
  dataUrl: string,
): Promise<string> {
  const apiKey = decryptApiKey(encryptedApiKey);
  if (!apiKey) throw new Error("No OpenRouter API key");
  const { text, inputTokens, outputTokens, costUsd, cachedTokens, cacheWriteTokens } =
    await visionCompletion(
      apiKey,
      CHARACTER_SYSTEM,
      [{ type: "text", text: "Describe this character portrait." }, imagePart(dataUrl)],
      2000,
    );
  await updateUsage(
    ownerDiscordId,
    inputTokens,
    outputTokens,
    costUsd,
    cachedTokens,
    cacheWriteTokens,
  );
  return text;
}

export async function describeKnownUserPortrait(
  ownerDiscordId: string,
  encryptedApiKey: string,
  dataUrl: string,
): Promise<string> {
  const apiKey = decryptApiKey(encryptedApiKey);
  if (!apiKey) throw new Error("No OpenRouter API key");
  const { text, inputTokens, outputTokens, costUsd, cachedTokens, cacheWriteTokens } =
    await visionCompletion(
      apiKey,
      KNOWN_USER_SYSTEM,
      [{ type: "text", text: "Describe this person's appearance." }, imagePart(dataUrl)],
      2000,
    );
  await updateUsage(
    ownerDiscordId,
    inputTokens,
    outputTokens,
    costUsd,
    cachedTokens,
    cacheWriteTokens,
  );
  let appearance = text;
  if (appearance && !appearance.toLowerCase().startsWith("is ")) {
    appearance = `is ${appearance.replace(/^\s+/, "")}`;
  }
  return appearance;
}

export function bufferToDataUrl(buf: Buffer, contentType: string): string {
  return `data:${contentType};base64,${buf.toString("base64")}`;
}
